// horizon-handler/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Events = exports.ComponentType = exports.CommandType = void 0;
// Commands Handler
var CommandType;
(function (CommandType) {
    CommandType[CommandType["ChatInput"] = 1] = "ChatInput";
    CommandType[CommandType["UserContext"] = 2] = "UserContext";
    CommandType[CommandType["MessageContext"] = 3] = "MessageContext";
})(CommandType || (exports.CommandType = CommandType = {}));
;
;
;
;
;
;
// Components Handler
var ComponentType;
(function (ComponentType) {
    ComponentType[ComponentType["Button"] = 1] = "Button";
    ComponentType[ComponentType["StringSelect"] = 2] = "StringSelect";
    ComponentType[ComponentType["UserSelect"] = 3] = "UserSelect";
    ComponentType[ComponentType["RoleSelect"] = 4] = "RoleSelect";
    ComponentType[ComponentType["MentionableSelect"] = 5] = "MentionableSelect";
    ComponentType[ComponentType["ChannelSelect"] = 6] = "ChannelSelect";
    ComponentType[ComponentType["Modal"] = 7] = "Modal";
})(ComponentType || (exports.ComponentType = ComponentType = {}));
;
;
;
;
;
;
;
;
;
;
;
var Events;
(function (Events) {
    Events["DeployStart"] = "deployStart";
    Events["DeployFinish"] = "deployFinish";
    Events["DeployError"] = "deployError";
    Events["FileSkip"] = "fileSkip";
    Events["FileLoad"] = "fileLoad";
})(Events || (exports.Events = Events = {}));
;
;
